### PR TITLE
Use queued & batched ReportStream uploader

### LINKED ISSUE #1312: Enable test event reporting queue worker
In order to enable the queue worker, we must:

1. Verify that the storage queue is being properly provisioned by the terraform work in #1301
2. Verify that an Azure Function is being created per-environment based on the work in #1425.
    - If it is **not**, we will have to configure this infrastructure
3. Add the storage queue application config to the Spring application, such that TestEvents start being published to the queue
4. Verify in `test` that the queue worker is reporting tests to ReportStream's test environment

From the [original doc](https://docs.google.com/document/d/129PINU3xTXDzZ5q4lUZdmlDUGOC_WUY0-fcii-9kino/edit#):

1. The first deployment would begin publishing serialized TestEvents to the Azure storage queue. 
    1. This deployment MUST NOT configure any trigger for the queue worker in order that these events not be published to the Data Hub via the queue-based channel.
    1. This deployment MUST NOT configure a dead letter queue or queue redrive policy.
1. As a last step of the scheduled job publishing a range of TestEvents to the Data Hub, the job should pull messages from the queue and “complete” any whose timestamp is less than or equal to the greatest processed timestamp captured as a cursor by the scheduled job. Messages with a timestamp greater than the cursor should be ignored (which will cause them to reappear in the queue).

### LINKED ISSUE #2126: Update Queue-Based Uploader to 100/batch Throttled Schedule
Our existing Azure Function code for queue-based uploads is configured to trigger on every write to the Storage Queue. It reads and uploads to ReportStream exactly one message from the queue on each execution.

In order to support ReportStream's individual upload throughput limitations, and to still migrate to our new queue-based upload infrastructure, we need need to change our implementation.

## Acceptance Criteria
- [x] The Azure Function for upload should not trigger when items are added to the queue
- [x] The function should instead execute on a schedule – for now, it should be the same "every 2hrs during the business day" that the SimpleReport app uses
- [x] This schedule is configurable, such that we can update it when ReportStream's throughput threshold improves
- [x] The new Azure function for uploads should:
    * Check that there are at least 100 items to read on the queue
    * Read up to 1000 items from the queue
    * Transform the JSON to CSV
    * Upload the items to ReportStream
    * Mark the queue items for deletion when upload succeeds

### PR DESCRIPTION
**Pull request by nullflux**
## Related Issue or Background Info

- Resolves #2126 
- Resolves #1312

## Changes Proposed

- Delete original queue-trigger uploader
- Disable scheduled uploads in the Spring app
- Add a Typescript Azure Function App that runs on a `5/15` cron Timer to:
    - Read from the queue
    - Compile a CSV
    - Upload it to ReportStream
  
## Additional Information

- Most of the lines (>10k) of this PR are `package-lock.json`! Don't be scared! 
- Since the queue has already been provisioned & the Spring app is already sending to it, this PR finishes prep for the migration
    - To finish the migration, all that's needed is a manual `terraform apply` for the function in prod
- The new function has to be built, compiled, and then uploaded
- The zip file is _also_ stored in source control

### UI
- [x] Any changes to the UI/UX are approved by design 
- [x] Any new or updated content (e.g. error messages) are approved by design 

### Testing
- [x] Includes a summary of what a code reviewer should verify

### Changes are Backwards Compatible
- [x] Database changes are submitted as a separate PR
  - [x] Any new tables that do not contain PII are accompanied by a GRANT SELECT to the no-PHI user
  - [x] Any changes to tables that have custom no-PHI views are accompanied by changes to those views
        (including re-granting permission to the no-PHI user if need be)
  - [x] Liquibase rollback has been tested locally using `./gradlew liquibaseRollbackSQL` or `liquibaseRollback`
  - [x] Each new changeset has a corresponding [tag](https://docs.liquibase.com/change-types/community/tag-database.html)
- [x] GraphQL schema changes are backward compatible with older version of the front-end

### Security
- [x] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [x] Any dependencies introduced have been vetted and discussed

## Cloud
- [x] DevOps team has been notified if PR requires ops support
- [x] If there are changes that cannot be tested locally, this has been deployed to our Azure `test`, `dev`, or `pentest` environment for verification

### CODE REVIEW

File: ops/services/app_functions/report_stream_batched_publisher/.gitignore [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# Terraform lockfile
+.terraform.lock.hcl
```
[the remark this answers is not in the document]
Comment (nullflux) [reply ops/services/app_functions/report_stream_batched_publisher/.gitignore:2]: Ah okay, thank you!

File: ops/services/app_functions/report_stream_batched_publisher/functions/.funcignore [hunk 2]
```diff
@@ -0,0 +1,7 @@
+*.js.map
+*.ts
+.git*
+.vscode
+local.settings.json
+test
+tsconfig.json
```
Comment (JSkeets) [inline ops/services/app_functions/report_stream_batched_publisher/functions/.funcignore:7]: nitpick: Newline

File: ops/services/app_functions/report_stream_batched_publisher/.gitignore (hunk 1 above)
[the remark this answers is not in the document]
Comment (nullflux) [reply ops/services/app_functions/report_stream_batched_publisher/.gitignore]: Resolved in ea6c873

File: ops/services/app_functions/report_stream_batched_publisher/functions/jest.config.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
+module.exports = {
+  preset: 'ts-jest',
+  testEnvironment: 'node',
+};
```
Comment (JSkeets) [inline ops/services/app_functions/report_stream_batched_publisher/functions/jest.config.js:5]: nitpick: Newline

File: ops/services/app_functions/report_stream_batched_publisher/functions/host.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+  "version": "2.0",
```
Comment (JSkeets) [inline ops/services/app_functions/report_stream_batched_publisher/functions/host.json:2]: Curious if this version here relates at all to the azure version in the package.json ? 

`version
This value indicates the schema version of host.json. The version string "version": "2.0" is required for a function app that targets the v2 runtime, or a later version. There are no host.json schema changes between v2 and v3.`

Although it states there are no differences between v2 and v3 it seems it would be best practice ? I could be off here thouse

Comment (JSkeets) [reply ops/services/app_functions/report_stream_batched_publisher/functions/host.json:2]: Ah it appears only options are version 2 and version 1 https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json

File: ops/services/app_functions/report_stream_batched_publisher/functions/proxies.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
+  "$schema": "http://json.schemastore.org/proxies",
+  "proxies": {}
+}
```
Comment (JSkeets) [inline ops/services/app_functions/report_stream_batched_publisher/functions/proxies.json:4]: Is this file necessary ?

File: ops/services/app_functions/report_stream_batched_publisher/_data.tf [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+data "azurerm_storage_account" "app" {
+  name                = "simplereport${var.environment}app"
+  resource_group_name = local.resource_group_name
+}
+
+data "azurerm_storage_account_sas" "sas" {
+  connection_string = data.azurerm_storage_account.app.primary_connection_string
+  https_only        = true
+  start             = "2021-09-01"
+  expiry            = "2022-12-31"
```
Comment (JSkeets) [inline ops/services/app_functions/report_stream_batched_publisher/_data.tf:32]: How are we not going to forget about this ? More importantly if the team is entirely new people in 1.25 years, how are they going to know about this expiration date ?

Comment (JSkeets) [reply ops/services/app_functions/report_stream_batched_publisher/_data.tf:32]: One idea is a slack reminder

File: ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/lib.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+const {
+  REPORT_STREAM_BATCH_MINIMUM,
+  REPORT_STREAM_BATCH_MAXIMUM,
+  REPORT_STREAM_TOKEN,
+  REPORT_STREAM_URL,
+  AZ_STORAGE_ACCOUNT_KEY,
+  AZ_STORAGE_ACCOUNT_NAME,
+  AZ_STORAGE_QUEUE_SVC_URL,
+  TEST_EVENT_QUEUE_NAME,
+} = ENV;
```
Comment (JSkeets) [inline ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/lib.ts:20]: I really appreciate all these variables declared here, makes this very flexible and maintainable, in my opinion.

File: ops/services/app_functions/report_stream_batched_publisher/functions/package.json [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  "name": "report-stream-batched-publisher",
+  "version": "1.0.0",
+  "description": "",
+  "scripts": {
+    "func": "func",
+    "clean": "rimraf build",
+    "build:compile": "tsc",
+    "build:production": "jest --coverage && npm run prestart && npm prune --production && npm run build:zip",
+    "prebuild:zip": "mkdir -m700 -p build",
+    "build:zip": "zip -r build/batched-rs-publisher.zip . -x \"*.zip\" -x \".terraform\" -x \"terraform.*\" -x \"*.tf\" -x \"local.settings.json\" -x \"*.map\"",
```
Comment (JSkeets) [inline ops/services/app_functions/report_stream_batched_publisher/functions/package.json:11]: There is a really cool package you can use to do this https://stuk.github.io/jszip/ :)

File: ops/services/app_functions/report_stream_batched_publisher/functions/host.json (hunk 4 above)
Comment (nullflux) [reply ops/services/app_functions/report_stream_batched_publisher/functions/host.json:2]: Ah, thanks for checking!

File: ops/services/app_functions/report_stream_batched_publisher/functions/proxies.json (hunk 5 above)
Comment (nullflux) [reply ops/services/app_functions/report_stream_batched_publisher/functions/proxies.json:4]: It was generated by the Azure Function App VSCode plugin when I created the project & function - I'd rather leave it than delete it and get burned

File: ops/services/app_functions/report_stream_batched_publisher/functions/package.json (hunk 8 above)
Comment (nullflux) [reply ops/services/app_functions/report_stream_batched_publisher/functions/package.json:11]: 😬 

So, we _started_ with https://github.com/ffflorian/jszip-cli - but it's not v actively maintained and I ran into a huge issue where directories that were too-deeply nested got excluded from the zip regardless of configuration and w/o any warning. This of course broke the build on some sub-dependencies

I switched to using system-native zip & mkdir (instead of jszip and mkdirp) to avoid this problem... of course it's no longer OS-agnostic, but 99% of this team is on Mac OS or Linux anyway. I'd like to keep this as-is for now and deal with it later if/when we have a Windows user who gets stuck

File: ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/lib.ts (hunk 7 above)
Comment (nullflux) [reply ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/lib.ts:20]: Thanks!

File: ops/services/app_functions/report_stream_batched_publisher/functions/.funcignore (hunk 2 above)
Comment (nullflux) [reply ops/services/app_functions/report_stream_batched_publisher/functions/.funcignore]: Resolved in d2f80ebe

File: ops/services/app_functions/report_stream_batched_publisher/functions/jest.config.js (hunk 3 above)
Comment (nullflux) [reply ops/services/app_functions/report_stream_batched_publisher/functions/jest.config.js]: Resolved in d2f80ebe

**Review by timothybest-usds — CHANGES_REQUESTED** [review]
Can you add the function paths to sonar coverage tracking in https://github.com/CDCgov/prime-simplereport/blob/main/backend/build.gradle#L187

the tests will probably need to be run in https://github.com/CDCgov/prime-simplereport/blob/main/.github/workflows/test.yml so that they are uploaded via gradle

File: ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+import { ReportStreamResponse } from "./rs-response";
+
+const { REPORT_STREAM_URL } = ENV;
+
+appInsights.setup();
+const telemetry = appInsights.defaultClient;
+
+const QueueBatchedTestEventPublisher: AzureFunction = async function (
+  context: Context,
+  myTimer: any
```
[the remark this answers is not in the document]
Comment (nullflux) [reply ops/services/app_functions/report_stream_batched_publisher/functions/QueueBatchedReportStreamUploader/index.ts]: The `any` actually came directly from Azure's example code for Timer-triggered functions 🤦 

I've removed it per your suggestion in 12792fab

**Comment by nullflux** [comment]
@timothybest-usds Integrated w/ Sonar in #2effae8. I'll eek out a little more coverage on `lib.ts` to round it out :)

**Review by timothybest-usds — APPROVED** [review]

File: ops/services/app_functions/report_stream_batched_publisher/functions/package.json (hunk 8 above)
Comment (JSkeets) [reply ops/services/app_functions/report_stream_batched_publisher/functions/package.json:12]: It was a bad joke, I know how much of a problem it caused you.

**Review by JSkeets — APPROVED** [review]
This is awesome! I know this was ton of work, and involved a lot of troubleshooting. I'd like to see a slack reminder for the SAS expiration date for future team members. Great work!

File: ops/services/app_functions/report_stream_batched_publisher/functions/package.json (hunk 8 above)
Comment (nullflux) [reply ops/services/app_functions/report_stream_batched_publisher/functions/package.json:12]: lolol 😭

File: ops/services/app_functions/report_stream_batched_publisher/_data.tf (hunk 6 above)
Comment (nullflux) [reply ops/services/app_functions/report_stream_batched_publisher/_data.tf:32]: Created! 

image
